csharp-advanced-topics/ActionFuncDelegatesCsharp/ActionFuncDelegatesCsharp/Program.cs

```diff
@@ -0,0 +1,65 @@
+﻿using System;
+
+namespace ActionFuncDelegatesCsharp
+{
+    public delegate string DelegateFullName(string firstName, string lastName);
+    public delegate void DelegatePrintSumOfNumbers(int firstNumber, int secondNumber);
+    public delegate bool DelegateCheckLengthOfString(string exampleString);
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            var firstNumber = 5;
+            var secondNumber = 10;
+            var exampleString = "AmazingCodeMaze";
+            var firstName = "Code";
+            var lastName = "Maze";
+
+            UnderstandingDelegates understandingDelegates = new UnderstandingDelegates();
+
+            #region Implementing Delegates Without Generics
+
+            Console.WriteLine("-------------- Prints Using Normal Delegate Invocation ---------------------");
+
+            DelegateFullName delegateFullName = understandingDelegates.GetFullName;
+            var fullName = delegateFullName.Invoke(firstName, lastName);
+            Console.WriteLine(fullName);
+
+            DelegatePrintSumOfNumbers delegateSum = understandingDelegates.PrintSumOfNumbers;
+            delegateSum.Invoke(firstNumber, secondNumber);
+
+            DelegateCheckLengthOfString delegateCheckLengthOfString = understandingDelegates.CheckLengthOfString;
+            Console.WriteLine(delegateCheckLengthOfString.Invoke(exampleString));
+
+            #endregion
+
+
+            #region Implementing Delegates With Generics
+            //Here we don't need to define delegates mentioned in line numbers 5,6 and 7
+            Console.WriteLine("\n-------------- Prints Using Generic Delegate Invocation --------------------\n");
+
+            Console.WriteLine("-------------- Implementing Func Delegate ----------------------------------");
+            Func<string, string, string> funcDelegateFullName = understandingDelegates.GetFullName;
```

Comment:
A good example on where we SHOULD use types because it's important to show Func works.